flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinitions.java

```diff
@@ -231,6 +232,21 @@ ANY, and(logical(LogicalTypeRoot.BOOLEAN), LITERAL)
                             "org.apache.flink.table.runtime.functions.scalar.ArrayContainsFunction")
                     .build();
 
+    public static final BuiltInFunctionDefinition ARRAY_SORT =
+            BuiltInFunctionDefinition.newBuilder()
+                    .name("ARRAY_SORT")
+                    .kind(SCALAR)
+                    .inputTypeStrategy(
+                            or(
+                                    sequence(new ArrayComparableElementArgumentTypeStrategy()),
+                                    sequence(
+                                            new ArrayComparableElementArgumentTypeStrategy(),
+                                            logical(LogicalTypeRoot.BOOLEAN))))
```

Comment:
Hi, @dawidwys , from the description, if second argument is null, we will return NUll.
`SELECT ARRAY_SLICE(['a', 'b', 'c', 'd', 'e'], null)
null
`

Reply:
You should update the output strategy then to take into account the second argument. So far you have only: `nullableIfArgs(argument(0))`

Reply:
Can you check what's the behaviour in other RDBMS? E.g. postgres, oracle, sql server?

Personally I find the behaviour a bit weird we accept nullable type for the second argument.

Reply:
Ok, I will check other RDNMS.

Reply:
If you're looking for ways to verify other databases that's a good start: https://sqlfiddle.com/

Reply:
Ok, thank you, actually, I did this function last summer and follow the jira tickect's https://issues.apache.org/jira/browse/FLINK-26948 description.
And I will check other RDNMS today.

Reply:
For postgresql https://www.postgresql.org/docs/8.4/intarray.html
it has three ways to sort an array
```
sort(int[], text dir) | int[] | sort array — dir must be asc or desc | sort('{1,2,3}'::int[], 'desc') | {3,2,1}
sort(int[]) | int[] | sort in ascending order | sort(array[11,77,44]) | {11,44,77}
sort_asc(int[]) | int[] | sort in ascending order |   |  
sort_desc(int[]) | int[] | sort in descending order |  
```

Reply:
> Ok, thank you, actually, I did this function last summer and follow the jira tickect's https://issues.apache.org/jira/browse/FLINK-26948 description.

I understand that. If at some point we find out we either made mistake or did not put enough effort into something it's better to fix that sooner rather than later when we need to leave with the consequences. I admit I have not thoroughly checked the semantics before which I should've.

It's better to do something well rather than fast in my opinion.

I see, so traditional RDBMS do not really support that function. It's also worth checking what does:
* Snowflake: https://docs.snowflake.com/en/sql-reference/functions/array_sort (`null` on any argument `null`), null handling separately
* Spark: https://docs.databricks.com/en/sql/language-manual/functions/sort_array.html: from the docs it does not say what's the behaviour on `null` `ascendingOrder`
* Presto: https://prestodb.io/docs/current/functions/array.html: has two separate functions for `ASC/DESC`

To me Snowflake's behaviour is the cleanest out there. WDYT? @MartijnVisser

Reply:
Ok, waiting for @MartijnVisser reply.

Reply:
This is quite a rabbit hole 😆 

1. Databricks has two functions:

* sort_array https://docs.databricks.com/en/sql/language-manual/functions/sort_array.html - Sorts ascending, puts NULL at beginning.
*  array_sort https://docs.databricks.com/en/sql/language-manual/functions/array_sort.html - Sorts ascending, puts NULL at the end. 

2. Spark has https://spark.apache.org/docs/latest/api/sql/index.html#sort_array which is indeed not different from the Databricks one. Sorts ascending, puts NULL at beginning

3. Snowflake indeed has https://docs.snowflake.com/en/sql-reference/functions/array_sort - Sorts ascending, puts NULL at the end. 

4. Trino, that has `array_sort(x)` - Sorts ascending, puts NULL at the end. 

5. StarRocks also has this function https://docs.starrocks.io/docs/2.2/sql-reference/sql-functions/array-functions/array_sort/ - Sorts ascending, puts NULL at beginning

6. Doris has it https://doris.apache.org/docs/1.2/sql-manual/sql-functions/array-functions/array_sort/ - Sorts ascending, puts NULL at beginning

7. Hive uses NULLS FIRST as the default behavior in ascending ordering mode, per https://issues.apache.org/jira/browse/HIVE-12994 - Sorts ascending, puts NULL at beginning.

I do think that the Snowflake has the cleanest function signature: there's just one function, and it has two optional arguments to give all the flexibility. 

I think the default sorting mode should be ascending, and then it boils down to decide on the default location on where to put NULL. Out of the 7 implementations, we have 5 who put NULL at the beginning, and 2 who put them at the end. Naturally, I think we should put NULL at the beginning but I'm +0 on the default location of NULL.

Reply:
@MartijnVisser Actually Snowflake does:
> nulls_first
 Default: FALSE if the ARRAY is sorted in ascending order; TRUE if the ARRAY is sorted in descending order.

I'd suggest we:
 
1. adapt Snowflake's signature where we get `ARRAY<COMPARABLE>, BOOLEAN, BOOLEAN`, output is `ARRAY<COMPARABLE>` which is `NULLABLE` if any of the input is nullable
2. We use defaults: ascending and `nulls_first` for ASC, `nulls_last` for DESC
3. If any of the input is null the output is also null

How does this sound?

Reply:
Sounds good to me, thanks!

Reply:
The updated function signature is now ARRAY_SORT(<array>[, <sort_ascending>[, <nulls_first>]]), where both sort_ascending and nulls_first parameters are optional, allowing for more flexibility in array sorting behavior.
@MartijnVisser @dawidwys

Reply:
Great description! 👍